### PR TITLE
fix(forge): properly handle `forge script` initialization errors

### LINKED ISSUE #2770: Forge Script Bug
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

0.2.0

### What command(s) is the bug in?

forge script

### Operating System

_No response_

### Describe the bug

When running a script, i get the following error:

The application panicked (crashed).
Message:  Function signature not found in the ABI
Location: cli/src/cmd/forge/script/mod.rs:411

### PR DESCRIPTION
**Pull request by joshieDo**
## Motivation
Long time coming. By default we call `run()` on `forge script`, and we expect the script contract to have no arguments. However, we were not properly handling when that didn't happen.

closes #2770
ref #2432

## Solution
Give a proper error when:
* the run signature is not found
* the constructor has arguments
* the constructor fails its execution

### CODE REVIEW

File: cli/src/cmd/forge/script/mod.rs [hunk 1]
```diff
@@ -408,7 +409,10 @@ impl ScriptArgs {
                 (
                     abi.functions()
                         .find(|&abi_func| abi_func.short_signature() == func.short_signature())
-                        .expect("Function signature not found in the ABI"),
+                        .wrap_err(format!(
+                            "Function `{}` is not implemented in your script.",
+                            self.sig
+                        ))?,
```
Comment (mattsse) [inline cli/src/cmd/forge/script/mod.rs:415]: that's the abi signature right?

File: cli/src/cmd/forge/script/runner.rs [hunk 2]
```diff
@@ -72,7 +72,11 @@ impl ScriptRunner {
             traces: constructor_traces,
             debug: constructor_debug,
             ..
-        } = self.executor.deploy(CALLER, code.0, 0u32.into(), None).expect("couldn't deploy");
+        } = self
+            .executor
+            .deploy(CALLER, code.0, 0u32.into(), None)
+            .map_err(|err| eyre::eyre!("Constructor failed:\n{}", err))?;
```
Comment (mattsse) [inline cli/src/cmd/forge/script/runner.rs:78]: "Failed to deploy contract"?

File: cli/src/cmd/utils.rs [hunk 3]
```diff
@@ -151,6 +151,16 @@ impl From<RetryArgs> for Retry {
     }
 }
 
+/// Returns error if constructor has arguments.
+pub fn ensure_clean_constructor(abi: &Abi) -> eyre::Result<()> {
+    if let Some(constructor) = &abi.constructor {
+        if !constructor.inputs.is_empty() {
+            eyre::bail!("Contract constructor should have no arguments. Add those arguments to  `run(...)` instead, and call it with `--sig run(...)`.");
```
Comment (mattsse) [inline cli/src/cmd/utils.rs:158]: "Constructor arguments are unsupported. Use `run()` function instead and provide the `--sig`... argument"

**Review by mattsse — CHANGES_REQUESTED** [review]
wording

File: cli/src/cmd/forge/script/runner.rs (hunk 2 above)
Comment (joshieDo) [reply cli/src/cmd/forge/script/runner.rs:78]: should allude to the constructor coming from the script contract

File: cli/src/cmd/utils.rs (hunk 3 above)
Comment (joshieDo) [reply cli/src/cmd/utils.rs:158]: afaik, people who put arguments on the constructor want to be able to call a script with arguments. This error message doesn't seem to help in that regard. They might think it's unsupported to do `run(uint a)`, which is not.

Comment (mattsse) [reply cli/src/cmd/utils.rs:158]: I see, totally agree, sgtm!

File: cli/src/cmd/forge/script/runner.rs (hunk 2 above)
Comment (mattsse) [reply cli/src/cmd/forge/script/runner.rs:78]: yeh good idea to add info that this is a script issue

**Review by mattsse — APPROVED** [review]

**Pull request merged**